### PR TITLE
[Cephadm][Tier-1]Automate cephadm bootstrap with docker

### PR DESCRIPTION
**Pull request by pranavprakash20**
# Description
Tests Included :

https://polarion.engineering.redhat.com/polarion/redirect/project/CEPH/workitem/question_mark/id=CEPH-83575202
https://polarion.engineering.redhat.com/polarion/redirect/project/CEPH/workitem/question_mark/id=CEPH-83575217

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

**Comment by pranavprakash20** [comment]
Run log : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-3XDJ9F/

File: tests/cephadm/test_cephadm_ansible_wrapper.py [hunk 1]
```diff
@@ -109,6 +109,12 @@ def run(ceph_cluster, **kwargs):
     module = aw.get("module")
     module_args = aw.get("module_args", {})
 
+    # Check if the scenario includes docker
+    if module_args.get("docker"):
```
Comment (vamahaja) [inline tests/cephadm/test_cephadm_ansible_wrapper.py:113]: This module arg is specific to bootstrap module. Better to move under `cephadm_bootstrap` check

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: tests/cephadm/test_cephadm_ansible_wrapper.py (hunk 1 above)
Comment (pranavprakash20) [reply tests/cephadm/test_cephadm_ansible_wrapper.py:113]: Its not only for bootstrap Vaibhav, the other scenario for logging into the registry via docker is also there. So it cannot be clubbed under bootstrap alone

**Comment by pranavprakash20** [comment]
Complete run results: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-HY7DA3/

**Comment by pranavprakash20** [comment]
run log of existing tests : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-3XDJ9F/

(adding it as a small change was made to bootstrap-cluster playbook)

File: cli/utilities/packages.py [hunk 2]
```diff
@@ -62,11 +62,55 @@ def install(self, pkg, nogpgcheck=False):
...
             pkg (str): package need to be installed
         """
-        cmd = f"{self.manager} install -y {pkg}"
-        if nogpgcheck:
-            cmd += " --nogpgcheck"
-        if self.execute(sudo=True, long_running=True, cmd=cmd):
-            raise PackageError(f"Failed to install package '{pkg}'")
+        # If the package to be installed is docker, follow
+        # the steps in https://docs.docker.com/engine/install/rhel/
+        if pkg == "docker":
```
Comment (vamahaja) [inline cli/utilities/packages.py:67]: These steps are not recommended. Installation guide doesn't mention any such process to install docker.

**Review by vamahaja — CHANGES_REQUESTED** [review]

**Comment by pranavprakash20** [comment]
New run log : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-IKAN1C/

The bootstrap tests fails due to the issue mentioned above. The test works when we install docker via other repos and when installing using the redhat repo, the bootstrap fails. Looks like this issue has to be raised.

File: cli/utilities/packages.py [hunk 3]
```diff
@@ -101,6 +115,16 @@ def clean(self):
         if self.execute(sudo=True, long_running=True, cmd=cmd):
             raise PackageError("Failed to clean repositories")
 
+    @staticmethod
```
Comment (vamahaja) [inline cli/utilities/packages.py:118]: This method can be utilised in many places, better to move it some common place.

File: cli/utilities/packages.py [hunk 4]
```diff
@@ -81,7 +88,14 @@ def upgrade(self, pkg):
...
         cmd = f"{self.manager} upgrade -y {pkg}"
-        if self.execute(sudo=True, long_running=True, cmd=cmd):
+
+        # When multiple nodes are passed, the execute returns a dict with return value
+        # for each node Each of these return values has to be checked to ensure that
+        # the package is upgraded in the specified nodes
+        out = self.execute(sudo=True, long_running=True, cmd=cmd)
+        if isinstance(out, dict):
+            self._verify_execution_status(out, pkg)
+        elif self.execute(sudo=True, long_running=True, cmd=cmd):
```
Comment (vamahaja) [inline cli/utilities/packages.py:98]: Why re-executing command ?

File: suites/pacific/cephadm/test_validate_cephadm_bootstrap_cases.yaml [hunk 5]
```diff
@@ -95,3 +95,9 @@ tests:
           playbook: playbooks/bootstrap-cluster-overide-exisitng-keys.yml
           module_args:
             mon_node: node1
+
```
Comment (vamahaja) [inline suites/pacific/cephadm/test_validate_cephadm_bootstrap_cases.yaml:98]: As we are creating separate test suite file, this test not required now.

File: tests/cephadm/test_cephadm_ansible_wrapper.py (hunk 1 above)
Comment (vamahaja) [reply tests/cephadm/test_cephadm_ansible_wrapper.py:113]: ok

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: suites/pacific/cephadm/test_validate_cephadm_bootstrap_cases.yaml (hunk 5 above)
Comment (pranavprakash20) [reply suites/pacific/cephadm/test_validate_cephadm_bootstrap_cases.yaml:98]: Noted, will remove it

File: cli/utilities/packages.py (hunk 3 above)
Comment (pranavprakash20) [reply cli/utilities/packages.py:118]: will move to utils

File: cli/utilities/packages.py (hunk 4 above)
Comment (pranavprakash20) [reply cli/utilities/packages.py:98]: noted.

File: cli/utilities/utils.py [hunk 6]
```diff
@@ -214,3 +214,15 @@ def get_builds_by_rhbuild(rhbuild):
         return match.group(1, 2)
 
     return
+
+
+def verify_execution_status(out, pkg):
```
Comment (vamahaja) [inline cli/utilities/utils.py:219]: this should not be specific to pkg. Instead of pkg make it cmd

**Review by vamahaja — APPROVED** [review]
Approving this PR as failed test case needs more investigation. Code logic and test is automtated as expected.

**Pull request merged**